File: java/AlberoBinario.java




public abstract class AlberoBinario extends Albero
{


    public AlberoBinario()
    {
        super();
        
    }
    public AlberoBinario( AlberoBinario b)
    {
        super(b.getRadice().copy());
    }
    public AlberoBinario(nodo n)
    {
        super(n);
    }

    public binary getRadice()
    {
        return (binary)(super.getRadice());
    }
     private boolean alberiUgualiHelp(binary n1,binary n2)
    {
        if((n1==null && n2==null) || (n1.numeroFigli()==0 && n2.numeroFigli()==0))
            return true;
        if((n1==null && n2!=null) || ( n1!=null && n2==null ) || ( n1.numeroFigli()!=n2.numeroFigli() ))
            return false;
        boolean ris=true;
        if(n1.left()!=null)
            ris=alberiUgualiHelp(n1.left(),n2.left());
        if(ris && n1.right()!=null)
            ris=alberiUgualiHelp(n1.right(),n2.right());
        return ris;
    }
    public boolean alberiUguali(Albero a) 
    {
        return alberiUgualiHelp((binary)getRadice(),(binary)a.getRadice());
    }
    public static void main(String []args)
    {
    }

}

File: TreeKalk/alberobinario.cpp

#include "alberobinario.h"

AlberoBinario::AlberoBinario():Albero(){}
AlberoBinario::AlberoBinario(const AlberoBinario& b,QString q):Albero(b.getRadice()->copy(),q){}
AlberoBinario::AlberoBinario(nodo* n,const QString& q):Albero(n,q){}

AlberoBinario::~AlberoBinario()
{
    //cout << "~alberobinario()\n";
}

bool AlberoBinario::operator ==(const AlberoBinario& b)const
{
    return Albero::operator ==(b);
}
bool AlberoBinario::operator !=(const AlberoBinario& b)const
{
    return !(operator ==(b));
}
AlberoBinario& AlberoBinario::operator =(const AlberoBinario& b)
{
    if(this!=&b)
    {
        delete getRadice();
        setRoot(b.getRadice()->copy());
    }
    return *this;
}

binary* AlberoBinario::getRadice()const
{
    return dynamic_cast<binary*>(Albero::getRadice());
}


